### DISCUSSION TITLE
Linked CUDA but cannot use GPU

### DISCUSSION THREAD

**Discussion by TD-Sky** [ggml-org/llama.cpp#8046] [Q&A]
## Environment

- OS: WSL2 Archlinux
- Toolchains (running with nix-shell): gcc11, cuda11.8
- GPU: NVIDIA GeForce RTX 4060 Ti, driver 552.44

## Steps

1. Compile with CUDA stubs (I changed Makefile)
2. Prepare a chat LLM in `models/llama-7b`, rename to `ggml-model-q4_0.gguf`.
3. Run `./examples/chat.sh`

## Expected and actual

It should utilize GPU, but the CUDA statistics in tasks manager is 0%.

I also have ollama-cuda which linked CUDA statically. It will utilize GPU.

I have done the same thing in Linux, it doesn't work as well.

![image](https://github.com/ggerganov/llama.cpp/assets/44722470/e2158e80-d704-4fc9-8eb8-7d2c1bfaeb1a)

**Comment by RhinoDevel** [comment]
Do you have an output log from the beginning of llama-cli run?

**Reply by TD-Sky** [reply]
I found `ggml_cuda_init: failed to initialize CUDA: CUDA driver is a stub library`. But compilation would fail when linking CUDA directly.

**Reply by TD-Sky** [reply]
Updated to the latest commit and compiled successfully. But there are still `CUDA driver is a stub library` errors.

**Reply by RhinoDevel** [reply]
Seems as if the "real" CUDA library cannot be found. E.g. ChatGPT gives some detailed hints, if you just enter "CUDA driver is a stub library". :-)

**Comment by dspasyuk** [comment]
You need to specify -n-gpu-layers 35 for example for a typical &b model. Something like this: 

../llama.cpp/llama-cli --model models/Meta-Llama-3-8B-Instruct_Q5_K_S.gguf --n-gpu-layers 25 -cnv  --interactive-first  --simple-io  -b 512 -n -1 --ctx_size 0 --temp 0.3 --top_k 10 --multiline-input  --repeat_penalty 1.12 -t 6 -r "/n>" --log-disable

**Reply by TD-Sky** [reply]
It utilized GPU when running on Linux physical machine but not on WSL. The latter can't even compile as both of them have same environment.

**Reply by TD-Sky** [reply]
Is your system running in WSL2?

**Reply by TD-Sky** [reply]
```nix
with import <nixpkgs> {};
(mkShell.override { stdenv = gcc11Stdenv; }) {
  buildInputs = [
    cudaPackages_11.cudatoolkit
    cudaPackages_11.cudnn
  ];
  shellHook = ''
    export CUDA_PATH=${cudaPackages_11.cudatoolkit}
    export EXTRA_LDFLAGS="-L/lib"
    export EXTRA_CCFLAGS="-I/usr/include"
  '';
}
```
